### PR TITLE
Re-enable ECAL sim, digi, rechit validation for phase 2 workflows.

### PR DESCRIPTION
**Pull request by tanmaymudholkar**
#### PR description:

This PR enables the ECAL DQM validation (simhits, digis, rechits, clusters, DQM) In the new phase2 workflows. This validation was so far not run for the phase2 workflows.

With this PR, the DQMIO output should fill the plots needed in the DataLayouts/Ecal subdirectory.

#### PR validation:

To test this PR, we have examined the output of the sample workflow 20034.0:
`runTheMatrix.py --nThreads=10 --list=20034.0
`
In addition we have run the basic battery of tests:
`runTheMatrix.py -l limited -i all --ibeos
`
It appears there are errors in workflows 10224.0 and 25202.0, but examining the logs, these don't seem to be due to the changes introduced by this PR.

#### if this PR is a backport please specify the original PR:
Not a backport.
EDIT: A PR to CMSSW master has also been made, seen [here](https://github.com/cms-sw/cmssw/pull/27549).

### CODE REVIEW

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
please test
Indeed the crashes look unrelated

**Comment by jfernan2** [comment]
@tanmaymudholkar  I don't understand the crash but I suggest you rebase your changes to 11_0_X and create a new PR in that release since that's the version used for the Phase2 Validation mainly

**Comment by jfernan2** [comment]
@tanmaymudholkar any progress on this?ECAL is no entering the phase2 validations...

**Comment by tanmaymudholkar** [comment]
Hi @jfernan2 , this was actually somewhat hard to fix... maybe the problem was that the config builder copied the mixing module configuration used here into other workflows. To solve the problem I copied the configuration under a different name. The tests all seem to run fine now, could you please re-test?

Also following your suggestion I've made an equivalent PR for 11_0_X: https://github.com/cms-sw/cmssw/pull/27549 ; I needed the same fix for the mixing module.

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
please abort

**Comment by jfernan2** [comment]
please test workflow 20034.0

**Comment by jfernan2** [comment]
@tanmaymudholkar  the PR is not only adding 1911 plots related to ECAL in the Phase2 relvals but also 65 plots on EcalDigisV/SelectiveReadout for every MC workflow tested:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_6_X_2019-07-17-2300+2a0439/32690/dqm-histo-comparison-summary.html
Besides, in the case of the Phase2 workflows, the FED folder is also added, which I believe it makes sense only for DQM Online
Is all this expected?

**Comment by tanmaymudholkar** [comment]
@jfernan2 : all these plots are the usual ECAL DQM plots and looked at by experts as part of the validation; they should be in the pre-phase2 workflows (produced by the default offline DQM workflow). About the 65 plots in EcalDigisV/SelectiveReadout: these are actually not new, they were always produced but stored in the wrong place in the DQMIO output. (You can see that the +65 is always accompanied by a -65.) I just added [this](https://github.com/tanmaymudholkar/cmssw/commit/c6c20b334566f1517f0db96dc70913fc37820009#diff-841afb6af5f25afe540f53adc440718eR853) line, which makes the difference. I think its absence so far was a bug.

As for the FED folder, I'm not sure, but it seems to be produced in the default offline DQM [workflow](https://cmssdt.cern.ch/lxr/source/DQMOffline/Ecal/python/ecal_dqm_source_offline_cff.py) which was why I included it here. I'm currently checking with ECAL experts whether they regularly examine those plots during validation.

**Comment by tanmaymudholkar** [comment]
@jfernan2 : as an update, I got some feedback from ECAL DPG:

1. The FED plots were indeed filled in earlier relval workflows, for example [here](https://tinyurl.com/y6lx7quv) and monitored. We'd like to keep them in the PR.

2. The 65 plots that are now moved to the correct subfolder may be quite valuable to ECAL trigger experts in the future to monitor changes to trigger and readout.

**Comment by jfernan2** [comment]
Thanks @tanmaymudholkar  but the FED plots you mean are in a different folder EcalBarrel / FEDIntegrity also present with the changes from your PR, which is fine
I meant those in FED/FEDIntegrity which is a general folder for all subsystems, not just ECAL. This one is usally monitores on Online DQM and here is duplicating information with the one above in EcalBarrel

**Comment by tanmaymudholkar** [comment]
Hi @jfernan2 : Thanks for the clarification. I looked a little deeper into this.

First, this duplication is present in the older relval workflows as well. You can see it [here](https://tinyurl.com/y5lnmxzw). I checked with a vanilla version of `CMSSW_10_6_0_patch2` and the sample phase2 workflow `20034.0` and uploaded the results to a temporary server on `lxplus749`. Assuming you've tunneled `localhost:8081` to `lxplus749:8081`, you can see the plots present [here](https://tinyurl.com/y4onl458) even without my changes. Curiously the plots are created but don't seem to be filled with vanilla CMSSW, whereas with my changes they do seem to be filled, see for example my test server on `lxplus735` [here](https://tinyurl.com/yyorapt5).

Of course, this duplication seems to be unnecessary. If these top-level FED plots are always going to be available, then we would not object to removing the duplicates from the ECAL workspace. If you think this needs to be done, please let me know if I should make another PR or modify this one.

**Comment by tanmaymudholkar** [comment]
@jfernan2 Sorry, I just realized looking over my plots again that they are not exact duplicates: compare these on `lxplus735`: [top-level](https://tinyurl.com/y5cpzbe4) and [ecal-customized](https://tinyurl.com/y3vgsykm). The latter seems to have the correct number of entries (10 events * 36 ECAL Barrel supermodules), while the former seems to be buggy, both here and in previous relval outputs: considering this, I would suggest that it is a good idea to leave the ECAL plots in until the top-level ones in FED/FEDIntegrity can be fixed.

**Comment by jfernan2** [comment]
@tanmaymudholkar  that's it, the top-level ones should not be there, they were there in non Phase2 workflows before your PR, but you are introducing them now in Phase2 workflows too
Could you please remove them everywhere? Top level FED folder is an Online DQM one

**Comment by tanmaymudholkar** [comment]
@jfernan2 , sorry, I actually meant that even **without** this PR, these plots were present in the phase2 workflows. You can see [here](https://tinyurl.com/yxllfezx) for example.

Upon further inspection, these plots are produced by the plugin "DQMFEDIntegrityClient", which is in the sequence [dqmFEDIntegrityClient](https://cmssdt.cern.ch/lxr/source/DQMServices/Components/python/DQMFEDIntegrityClient_cff.py#0021).

This sequence is part of the harvesting step in phase 2. For example, using a vanilla CMSSW release and trying out phase2 workflow 20034.0:

```
[tmudholk@lxplus782 temp]$ runTheMatrix.py -l 20034.0 --dryRun | grep "step4"
 cmsDriver.py step4  --conditions auto:phase2_realistic -s HARVESTING:@phase2Validation+@phase2+@miniAODValidation+@miniAODDQM --scenario pp --filetype DQM --geometry Extended2023D17 --era Phase2 --mc  -n 100  --filein file:step3_inDQM.root --fileout file:step4.root  > step4_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D17_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D17+RecoFullGlobal_2023D17+HARVESTFullGlobal_2023D17.log  2>&1
[tmudholk@lxplus782 temp]$ cmsDriver.py step4 --no_exec --conditions auto:phase2_realistic -s HARVESTING:@phase2Validation+@phase2+@miniAODValidation+@miniAODDQM --scenario pp --filetype DQM --geometry Extended2023D17 --era Phase2 --mc  -n 100  --filein file:step3_inDQM.root --fileout file:step4.root
HARVESTING:@phase2Validation+@phase2+@miniAODValidation+@miniAODDQM
entry file:step3_inDQM.root
Step: HARVESTING Spec: ['@phase2Validation', '@phase2', '@miniAODValidation', '@miniAODDQM']
Config file step4_HARVESTING.py created
[tmudholk@lxplus782 temp]$ edmConfigDump step4_HARVESTING.py > step4_HARVESTING_dump.py
```
In `step4_HARVESTING_dump.py`, you can see (relevant extract):

```
...
process.dqmFEDIntegrity = cms.EDAnalyzer("DQMFEDIntegrityClient",
...
process.dqmFEDIntegrityClient = cms.Sequence(process.dqmFEDIntegrity+process.dqmQTestDQMFED)
...
process.DQMHarvestOuterTracker = cms.Sequence(process.dqmDcsInfoClient+process.OuterTrackerClient+process.dqmFEDIntegrityClient+process.DQMMessageLoggerClientSeq+process.dqmFastTimerServiceClient)
...
process.schedule = cms.Schedule(*[ process.postValidation_common_step, process.postValidationTracking_step, process.postValidation_muons_step, process.postValidation_JetMET_step, process.bTagCollectorSequenceMCbcl_step, process.postValidation_HCAL_step, process.hgcalValidatorPostProcessor_step, process.postValidationOuterTracker_step, process.DQMHarvestTracking_step, process.DQMHarvestOuterTracker_step, process.DQMHarvestMuon_step, process.DQMCertMuon_step, process.DQMHarvestHcal_step, process.HcalDQMOfflinePostProcessor_step, process.DQMHarvestEGamma_step, process.validationHarvestingMiniAOD, process.DQMHarvestMiniAOD_step, process.dqmsave_step ], tasks=[process.patAlgosToolsTask])
```

Conclusions:
1. The top-level plots in "FED/FEDIntegrity" are created by default in the phase2 workflows also, regardless of this PR. However, they are empty in the default workflows.
2. It is true that with this PR, these plots are now no longer empty. The difference made by this PR appears to be due to the way they are filled, see source code [here](https://cmssdt.cern.ch/lxr/source/DQMServices/Components/plugins/DQMFEDIntegrityClient.cc#0159); as you can see it just copies the number of entries from the individual subsystem plots.

One solution might be to ask the tracker DQM experts whether it's possible to remove the "dqmFEDIntegrityClient" module from their relval sequence.

**Comment by jfernan2** [comment]
+1
Thanks @tanmaymudholkar  It is clear now.
Please try to fix the other issue I sent you about warnings from ECAL in logs from Cosmics and other workflows

File: Configuration/StandardSequences/python/Validation_cff.py [hunk 1]
```diff
@@ -103,3 +103,23 @@
 from Configuration.Eras.Modifier_fastSim_cff import fastSim
 fastSim.toReplaceWith(validation_prod,_validation_prod_fastsim)
 
+from Validation.EcalClusters.ecalClustersValidationSequence_cff import ecalClustersValidationSequence
```
Comment (fabiocos) [inline Configuration/StandardSequences/python/Validation_cff.py]: @tanmaymudholkar @jfernan2 sorry, but I do not think this is the correct place where to put such a definition, why not in the specialized ECAL validation configuration https://github.com/cms-sw/cmssw/blob/master/Validation/Configuration/python/ecalSimValid_cff.py ?

Comment (jfernan2) [reply Configuration/StandardSequences/python/Validation_cff.py]: @fabiocos I do agree... and the same applies to the just merged master version of this #27549

**Comment by civanch** [comment]
+1

File: Configuration/StandardSequences/python/Validation_cff.py (hunk 1 above)
Comment (fabiocos) [reply Configuration/StandardSequences/python/Validation_cff.py]: @jfernan2 sorry, but there is nothing merged yet, I have put that on hold pending a clarification and code rearrangment

**Comment by fabiocos** [comment]
hold

**Comment by jfernan2** [comment]
-1
Until python configuration is moved to proper file

File: Configuration/StandardSequences/python/Validation_cff.py (hunk 1 above)
Comment (tanmaymudholkar) [reply Configuration/StandardSequences/python/Validation_cff.py]: @jfernan2 @fabiocos : Indeed the ECAL-specific configuration fragment seems like a better place for this definition. I've put it there and tested the workflows in the latest commit.

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by kpedro88** [comment]
Why is this PR needed in 10_6_X? Phase 2 development continues in 11_0_X.

**Comment by kpedro88** [comment]
backport of #27549

**Comment by tanmaymudholkar** [comment]
@kpedro88 : this issue was first noticed by ECAL DPG while examining a recent relval sample produced with 10_6_0_patch2. ECAL DPG needs the plots introduced by this PR as part of the validation process, so the idea is to include them in future 10_6_X releases.

**Comment by fabiocos** [comment]
@tanmaymudholkar reiterating the question of @kpedro88 : why in 10_6_X? What is the practical use case for code dedicated to validate developments of phase2 (to happen in 11_0_X) in a release mostly meant to analyze Run2 (apart for L1 TDR)?

**Comment by crovelli** [comment]
@fabiocos 
PdMV recently opened a few relval campaigns to validate the phaseII scenario in view of the L1 TDR. Ecal was basically blind because of these missing plots. That's why DPG asked Tanmay to do this work. If no more validation campaign for phaseII will come in 10.6.x we're happy to have this development only in 11_0_X. Otherwise, it would be useful to have everything in the 10.6.x serie as well. Thanks a lot.

@zghiche @fcouderc

**Comment by fabiocos** [comment]
@crovelli thank you for the clarification. @prebello @zhenhu are we expecting further Phase2 oriented validation campaigns in 10_6_X in the coming weeks?

**Comment by civanch** [comment]
+1

**Comment by silviodonato** [comment]
I'm closing the PRs with no activity in the last months.
@tanmaymudholkar are you still actually interested in this backport to 10_6_X (CMSSW_11_0_0 is out since December)? 
If not, I suggest to close this PR. Thanks!

**Comment by silviodonato** [comment]
If you want to re-open this PR, please let us know.

**Pull request closed without merging**